src/ClientModel/SamplingDataPostProcessor.cpp

```diff
@@ -170,47 +243,55 @@ void SamplingDataPostProcessor::SortByThreadUsage() {
-              exact_address_to_function_address_.end()) {
-            MapAddressToFunctionAddress(address, capture_data);
-          }
-          uint64_t function_address = exact_address_to_function_address_.at(address);
+  callstack_data.ForEachUniqueCallstack([this, &capture_data](uint64_t callstack_id,
+                                                              const CallstackInfo& callstack) {
+    // A "resolved callstack" is a callstack where every address is replaced by the start address of
+    // the function (if known).
+    std::vector<uint64_t> resolved_callstack_frames;
+
+    for (uint64_t address : callstack.frames()) {
+      if (exact_address_to_function_address_.find(address) ==
+          exact_address_to_function_address_.end()) {
+        MapAddressToFunctionAddress(address, capture_data);
+      }
+      uint64_t function_address = exact_address_to_function_address_.at(address);
+      resolved_callstack_frames.push_back(function_address);
+    }
 
-          resolved_callstack_frames.push_back(function_address);
-          function_address_to_sampled_callstack_ids_[function_address].insert(callstack_id);
-        }
+    if (callstack.type() == CallstackInfo::kComplete) {
+      for (uint64_t function_address : resolved_callstack_frames) {
+        function_address_to_sampled_callstack_ids_[function_address].insert(callstack_id);
+      }
+    } else {
+      // For non-kComplete callstacks, only use the innermost frame for statistics.
+      function_address_to_sampled_callstack_ids_[resolved_callstack_frames[0]].insert(callstack_id);
+    }
 
-        // Check if we already have this callstack
-        uint64_t resolved_callstack_id;
-        auto it = resolved_callstack_to_id_.find(resolved_callstack_frames);
-        if (it == resolved_callstack_to_id_.end()) {
-          resolved_callstack_id = callstack_id;
-          CHECK(!id_to_resolved_callstack_.contains(resolved_callstack_id));
-          orbit_client_protos::CallstackInfo resolved_callstack;
-          *resolved_callstack.mutable_frames() = {resolved_callstack_frames.begin(),
-                                                  resolved_callstack_frames.end()};
-          resolved_callstack.set_type(callstack.type());
-          id_to_resolved_callstack_.insert_or_assign(resolved_callstack_id, resolved_callstack);
-          resolved_callstack_to_id_.insert_or_assign(std::move(resolved_callstack_frames),
-                                                     resolved_callstack_id);
-        } else {
-          resolved_callstack_id = it->second;
-        }
+    CallstackInfo::CallstackType resolved_callstack_type = callstack.type();
+
+    // Check if we already have this resolved callstack, and if not, create one.
+    uint64_t resolved_callstack_id;
+    auto it = resolved_callstack_to_id_.find(CallstackInfoAsPairWithLvalueRefToFrames{
+        resolved_callstack_frames, resolved_callstack_type});
+    if (it == resolved_callstack_to_id_.end()) {
```

Comment:
`it` is used below and we don't want to query the map twice.

Reply:
My very personal point of view on this, is that the contains is much more readable, and the performance gain by not querying twice is not large enough to justify this. But as I said, this is very much personal preference, so feel free to keep those.

Reply:
Why is the `find`/`== end()` pattern not readable? It feels wrong to query twice when not necessary.

Reply:
I feel that it hides the intent (checking if the key is inside the map) and also exposing the iterator would not be necessary (that kind of goes with go/totw/158). 
However, go/totw/132 is quite clear about avoiding the second lookup, so I'll stop complaining.

Also that's said, I would probably be less opinionated about the find method if it would return an optional rather than the iterator...

Reply:
The intent is not only to check if the key is in the map, but also the get the value if it exists. That's why go/totw/158 doesn't apply here and why we need to do this.

> I would probably be less opinionated about the find method if it would return an optional rather than the iterator

The `find`/`== end()` has been there forever and is well known (in particular well before `optional` was added).